gslib/utils/wrapped_credentials.py

```diff
@@ -0,0 +1,166 @@
+import copy
+import datetime
+import json
+
+import google.auth.aws
+import google.auth.credentials
+import google.auth.identity_pool
+
+import oauth2client
+from google.auth.transport import requests
+from gslib.utils import constants
+
+# Expiry is stored in RFC3339 UTC format
+EXPIRY_FORMAT = '%Y-%m-%dT%H:%M:%SZ'
+
+DEFAULT_SCOPES = [
+    constants.Scopes.CLOUD_PLATFORM,
+    constants.Scopes.CLOUD_PLATFORM_READ_ONLY,
+    constants.Scopes.FULL_CONTROL,
+    constants.Scopes.READ_ONLY,
+    constants.Scopes.READ_WRITE,
+]
+
+
+class WrappedCredentials(oauth2client.client.OAuth2Credentials):
+
+  def __init__(self, base):
+    if not isinstance(base, google.auth.credentials.Credentials):
+      raise TypeError("Invalid Credentials")
+    self._base = base
+    super(WrappedCredentials, self).__init__(None, base._audience, None, None,
+                                             None, None, None)
+
+  def _do_refresh_request(self, http):
+    self._base.refresh(requests.Request())
+    if self.store is not None:
+      self.store.locked_put(self)
+
+  @property
+  def access_token(self):
+    return self._base.token
+
+  @access_token.setter
+  def access_token(self, value):
+    self._base.token = value
+
+  @property
+  def token_expiry(self):
+    return self._base.expiry
+
+  @token_expiry.setter
+  def token_expiry(self, value):
+    self._base.expiry = value
+
+  def _to_json(self, strip, to_serialize=None):
```

Comment:
This seems to be very similar to the implementation present in Oauth2client https://github.com/googleapis/oauth2client/blob/50d20532a748f18e53f7d24ccbe6647132c979a9/oauth2client/client.py#L241

Any reason why we don't just call super().to_json() and then update the serialized data?

```
def to_json(self):
  serialized_data = super().to_json()
  deserialized_data = json.loads(serialized_data)
  deserialized_data['_base'] = ...
  #. add/replace any other additional data.
  return json.dumps(deserialized_data)
```

This will be much shorter and also  avoid having the need to override a private method.

Reply:
Done

Reply:
I think we should override the `to_json` method instead of `_to_json`.

Reply:
Done